### PR TITLE
NOISSUE - Handle authorization errors other than unauthorize

### PR DESCRIPTION
**Pull request by dborovcanin**

### CODE REVIEW

File: ws/api/transport.go [hunk 1]
```diff
@@ -55,6 +57,18 @@ func handshake(svc ws.Service) http.HandlerFunc {
...
+				case codes.PermissionDenied:
+					w.WriteHeader(http.StatusUnauthorized)
+				case codes.InvalidArgument:
+					w.WriteHeader(http.StatusBadRequest)
+				default:
+					w.WriteHeader(http.StatusServiceUnavailable)
+				}
+				return
+			}
 			logger.Warn(fmt.Sprintf("Failed to authorize: %s", err))
```
Comment (anovakovic01) [inline ws/api/transport.go]: Log error before extracting error code.

File: ws/api/transport.go [hunk 2]
```diff
@@ -55,6 +57,18 @@ func handshake(svc ws.Service) http.HandlerFunc {
 			return
 		}
 		if err != nil {
+			e, ok := status.FromError(err)
+			if ok {
+				switch e.Code() {
+				case codes.PermissionDenied:
+					w.WriteHeader(http.StatusUnauthorized)
```
Comment (anovakovic01) [inline ws/api/transport.go]: Use forbidden instead.

File: ws/api/transport.go [hunk 3]
```diff
@@ -55,6 +57,18 @@ func handshake(svc ws.Service) http.HandlerFunc {
...
+				case codes.InvalidArgument:
+					w.WriteHeader(http.StatusBadRequest)
```
Comment (anovakovic01) [inline ws/api/transport.go]: This case is handled in `authorize` function. Request won't be sent if there is no token. If you want, you can update error there.

File: ws/api/transport.go [hunk 4]
```diff
@@ -55,6 +57,18 @@ func handshake(svc ws.Service) http.HandlerFunc {
...
+				default:
+					w.WriteHeader(http.StatusServiceUnavailable)
```
Comment (anovakovic01) [inline ws/api/transport.go]: Should client receive service unavailable in default case?

File: http/api/transport.go [hunk 5]
```diff
@@ -111,6 +111,18 @@ func encodeResponse(_ context.Context, w http.ResponseWriter, response interface
 }
 
 func encodeError(_ context.Context, err error, w http.ResponseWriter) {
+	e, ok := status.FromError(err)
+	if ok {
+		switch e.Code() {
+		case codes.PermissionDenied:
+			w.WriteHeader(http.StatusUnauthorized)
```
Comment (anovakovic01) [inline http/api/transport.go]: Use forbidden instead.

File: http/api/transport.go [hunk 6]
```diff
@@ -111,6 +111,18 @@ func encodeResponse(_ context.Context, w http.ResponseWriter, response interface
...
+		case codes.InvalidArgument:
+			w.WriteHeader(http.StatusBadRequest)
```
Comment (anovakovic01) [inline http/api/transport.go]: This case is handled in `authorize` function.

File: ws/api/transport.go [hunk 7]
```diff
@@ -55,7 +57,19 @@ func handshake(svc ws.Service) http.HandlerFunc {
 			return
 		}
 		if err != nil {
-			logger.Warn(fmt.Sprintf("Failed to authorize: %s", err))
+			e, ok := status.FromError(err)
+			if ok {
+				logger.Warn(fmt.Sprintf("Failed to authorize: %s", err))
+				switch e.Code() {
+				case codes.PermissionDenied:
+					w.WriteHeader(http.StatusUnauthorized)
```
Comment (anovakovic01) [inline ws/api/transport.go]: Use `forbidden` instead.

File: ws/api/transport.go [hunk 8]
```diff
@@ -55,7 +57,19 @@ func handshake(svc ws.Service) http.HandlerFunc {
 			return
 		}
 		if err != nil {
-			logger.Warn(fmt.Sprintf("Failed to authorize: %s", err))
+			e, ok := status.FromError(err)
+			if ok {
+				logger.Warn(fmt.Sprintf("Failed to authorize: %s", err))
```
Comment (anovakovic01) [inline ws/api/transport.go]: Log every error, not just those with code.

**Review by anovakovic01 — DISMISSED** [review]

File: clients/api/grpc/endpoint_test.go [hunk 9]
```diff
@@ -70,16 +68,19 @@ func TestCanAccess(t *testing.T) {
 		clientKey string
 		chanID    string
 		id        string
-		err       error
+		respCode  codes.Code
```
Comment (mijicd) [inline clients/api/grpc/endpoint_test.go]: Please rename to `code`.

File: clients/mocks/security.go [hunk 10]
```diff
@@ -24,7 +28,7 @@ func (hm *hasherMock) Compare(plain, hashed string) error {
 type identityProviderMock struct{}
 
 func (idp *identityProviderMock) TemporaryKey(id string) (string, error) {
-	if id == "" {
+	if id == "" || strings.Contains(id, "wrong") {
```
Comment (mijicd) [inline clients/mocks/security.go]: What's the purpose of this?

**Review by mijicd — CHANGES_REQUESTED** [review]

File: http/api/transport.go [hunk 11]
```diff
@@ -111,15 +111,28 @@ func encodeResponse(_ context.Context, w http.ResponseWriter, response interface
 }
 
 func encodeError(_ context.Context, err error, w http.ResponseWriter) {
+	e, ok := status.FromError(err)
```
Comment (mijicd) [inline http/api/transport.go]: Infline with if and move to the `default` case of the switch below.

File: ws/api/transport.go [hunk 12]
```diff
@@ -56,6 +58,18 @@ func handshake(svc ws.Service) http.HandlerFunc {
 		}
 		if err != nil {
 			logger.Warn(fmt.Sprintf("Failed to authorize: %s", err))
+			e, ok := status.FromError(err)
```
Comment (mijicd) [inline ws/api/transport.go]: Same here.

**Review by mijicd — CHANGES_REQUESTED** [review]

File: http/api/transport.go (hunk 11 above)
Comment (dusanb94) [reply http/api/transport.go]: Can you explain why we should remove `default` case? If auth server fails, default is an internal server error, which makes our service unavailable.

**Review by mijicd — APPROVED** [review]

**Pull request merged**